nixos/system/boot/initrd-openvpn: New openvpn options for initrd

CRTified opened the pull request:
The module in this commit adds new options that allows the integration of an OpenVPN client into the initrd.

###### Motivation for this change

The main motivation for this change is the ability to remotely unlock LUKS containers on hosts that are behind a restrictive firewall, but are able to connect to other hosts. Using a VPN connection allows that host to tunnel into a network that might be reachable by external devices (either by providing a public IPv4/6 address or by connecting into a private VPN).

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

peterhoeg commented on nixos/modules/system/boot/initrd-openvpn.nix:
```diff
@@ -0,0 +1,112 @@
+    '';
+
+    boot.initrd.network.postCommands = ''
+      openvpn /etc/openvpn/ovpn-initrd-cfg.ovpn &
```
Don't we have systemd in the initrd?

Amarandus replied:
I've only seen `systemd-udevd`, `libsystemd-shared-239.so` and `libsystemd.so.0` in the initramfs, and as `initrd-ssh.nix` uses the same approach to start `dropbear` ([see here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/initrd-ssh.nix#L122)), I assumed that this is the correct way of starting the client.
Additionally, this gives us the output of the client on the tty, which can make configuration/connection problems more obvious (even when e.g. restricting the shell to `/bin/cryptsetup-askpass`, what I guess most people do).

Mic92 replied:
No we don't

peterhoeg commented:
I like this - I don't have a need for it, but it looks like the right thing to do.

The only thing that concerns me is stuff that messes around with initrd without a test. Something basic would be fine as long as we are sure that enabling this doesn't break booting.

Amarandus commented:
I'll have a look into writing a test and will add it as soon as it is ready.

Amarandus commented:
Would it make sense to squash the two commits, as they both just add tests (and one commit consumes the other one)?

lheckemann commented:
This is cool! I'm still a bit -1 on this though. While we don't have any formal criteria for inclusion/exclusion of features, I think this is a use case specific and rare enough that it will only be used by you and might be better off in your personal config than in nixpkgs.

duhirsch commented:
I support this.

I think the usecase is not too niche.
Imagine the following usecase:
You want to run a homeserver with encrypted disks
Your ISP does not give you a public IP
Your homeserver reboots/looses power
You are currently not at home

Now you need a possibility to connect to the homeserver and unlock the disks.
Assuming you have another server which has a public IP address you can connect to it via VPN and use it as a jump server to enter the password

Lassulus commented:
I'm also a big fan of this PR, sadly I couldn't test it because of general busyness. Codewise it looks good to me.

peterhoeg commented:
@Amarandus, can you get someone else to test it as well (again due it messing with the boot process) to make sure all is good? Maybe a post on https://discourse.nixos.org if you don't have anyone at hand to try it out?

Mic92 commented on nixos/tests/initrd-network-openvpn/default.nix:
```diff
@@ -0,0 +1,150 @@
+  };
+
+
+  testScript =
```
This needs to be converted to python now.

CRTified replied:
I'll update it as soon as I have a bit of spare time :) But I still need to look for people willing to test...

Mic92 replied:
Looks like @Lassulus would test this.

Mic92 replied:
But since we have a test this looks actually good enough.

CRTified replied:
Testing is now ported to python.

CRTified committed:
nixos/system/boot/initrd-openvpn: Add openvpn options for initrd
nixos/tests/initrd-openvpn: Add test for openvpn in the initramfs

The module in this commit adds new options that allows the
integration of an OpenVPN client into the initrd.
This can be used e.g. to remotely unlock LUKS devices.

This commit also adds two tests for `boot.initrd.network.openvpn`.
The first one is a basic test to validate that a failing connection
does not prevent the machine from booting.

The second test validates that this module actually creates a valid
openvpn connection.
For this, it spawns three nodes:

  - The client that uses boot.initrd.network.openvpn
  - An OpenVPN server that acts as gateway and forwards a port
    to the client
  - A node that is external to the OpenVPN network

The client connects to the OpenVPN server and spawns a netcat instance
that echos a value to every client.
Afterwards, the external node checks if it receives this value over the
forwarded port on the OpenVPN gateway.

CRTified commented:
So, I finally got time to update this PR and squashed the commits together.

 - I removed most options as they were unnecessary (adding additional files like certificates should be possible with `boot.initrd.secrets` directly, as the `internal`-attribute was removed recently there. Alternatively, you can inline them into the configuration)
   - Besides `enable`, there is only the option `configuration`, which expects a file path as a string (similar to `boot.initrd.network.ssh.hostKeys`)
   - Configuration is now always handled by `boot.initrd.secrets`, so there is only one place which decides whether files get into the store or not (depends on the bootloader)
   - It is not expected anymore to add `iproute /bin/ip` to the config - it will be appended automatically
 - The test now uses the python test infrastructure
   - First test: Checks whether booting with an invalid config breaks
   - Second test: Checks whether a valid configuration really works
